### PR TITLE
Command handler

### PR DESCRIPTION
**Pull request by NiLSPACE**
**DO NOT MERGE**
Work in progress. Creating a file for each category, and moving the handlers to it.

### CODE REVIEW

**Comment by NiLSPACE** [comment]
Anyone know where I should put the command helper functions? (Commands/functions.lua)
I was thinking of merging it with the other functions file.

**Comment by NiLSPACE** [comment]
Any ideas?

**Comment by madmaxoft** [comment]
Can't those functions be in a top-level source file?

**Comment by NiLSPACE** [comment]
Yes, but we already have a file named functions there. Should I merge it with that?

**Comment by madmaxoft** [comment]
I have no opinion on this. Sorry.

**Comment by NiLSPACE** [comment]
So, anyone up to review the code, or are we just gonna merge it? Many of the line changes are just code being moved from one file to the other, but there are also a few things that I changed. For example I added the `cExpression` and `cShapeGenerator` class.

**Comment by NiLSPACE** [comment]
Oh, I'd like to make a change in the cExpression class. I'll do it tomorrow.

**Comment by NiLSPACE** [comment]
There, I'm done now :)

**Comment by madmaxoft** [comment]
I thought I wrote a commend, but can't find it anymore, so I guess I forgot to submit it, and closed the window :(

The classes need more documentation. I have no idea what `cExpression` does and how to use it. Likewise with `cShapeGenerator`.

**Comment by NiLSPACE** [comment]
Is this sufficient?

**Comment by madmaxoft** [comment]
I still don't quite understand what the bound parameters and variables in `cExpression` are.

**Comment by NiLSPACE** [comment]
The function cExpression:Compile() returns would look like this:
```lua
function ()
   -- All the formula stuff
   return
end
```

Using cExpression:BindParam you can add parameters to the function. If you then use 
```lua
cExpression:BindParam("data", true --[[Return the parameter]], true --[[It's also an argument ]])
```

The function would look like this:
```lua
function(data)
   -- The formula stuff
   return data
end
```

**Comment by NiLSPACE** [comment]
**Bump** ;)

**Comment by madmaxoft** [comment]
Is it wise, combining parameters with return values in this way?

**Comment by NiLSPACE** [comment]
I don't quite follow you. :|

**Comment by madmaxoft** [comment]
If I understood correctly, by using `BindParam`, you're saying either "here, take this value" or "give me that value as a return" (or both). I think that's rather poor interface, having two things within one function call.

**Comment by NiLSPACE** [comment]
Well, you basicly add a parameter, and you can say if it should be added as a parameter to the formula, and/or if it should be returned.

**Comment by NiLSPACE** [comment]
How would you suggest to do it otherwise? Have 2 functions? One for adding a return value, and one to add an argument to the formula?

**Comment by madmaxoft** [comment]
Yes, two separate functions would make much more sense to me.
And also this whole discussion means that the documentation is still lacking, so you should probably add to it.

Regarding those "variables", I still don't know what they are, but I have a feeling they are more like named constants, aren't they? How about renaming the API, then?

**Comment by NiLSPACE** [comment]
Okay, so I split up `BindParam` into `AddFormulaVariable` and `AddReturnValue`. I'm not sure if I like those names. Any suggestions or do you think they are fine?

I do think `PredefineConstant` would be better then `PredefineVariable`, so I'll rename that one.

**Comment by madmaxoft** [comment]
`AddParameter` would be a bit better name, in my opinion.

**Comment by NiLSPACE** [comment]
Done

**Comment by NiLSPACE** [comment]
No further objections?

File: Classes/Expression.lua [hunk 1]
```diff
@@ -0,0 +1,230 @@
+
+-- Expression.lua
+
+-- Contains the cExpression class. This allows formulas to be executed safely in an empty environment.
+
```
Comment (madmaxoft) [inline Classes/Expression.lua]: A quick "usage" comment would be nice here, to give an example.

File: Classes/Expression.lua [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	table.insert(self.m_PredefinedConstants, {name = a_VarName, value = a_Value})
+	return self
+end
+
+
+
+
+
+-- Creates a safe function from the formula string, the bound parameters and the predefined variables.
+function cExpression:Compile()
```
Comment (madmaxoft) [inline Classes/Expression.lua]: The returned function takes the previously-bound parameters (AddParameter()), does the calculations using any predefined constants (PredefineConstant()) and returns the named values (AddReturnValue())

File: Classes/Expression.lua [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	end
+	
+	-- The number of comparisons. This will be used to give each comparison a name (Comp<nr>)
+	local NumComparison = 1
+	
+	-- Split the formula into actions (For example in "data=5; x<y" data=5 is an action, and x<y is an action.)
+	local Actions = StringSplitAndTrim(self.m_Formula, ";")
+	
+	-- Loop through each action to check if the action is an comparison or an assignment. Handle the actions accordingly.
+	for Idx, Action in ipairs(Actions) do
```
Comment (madmaxoft) [inline Classes/Expression.lua]: WHY do we need to handle actions differently? That should be in the comment, not WHAT we're doing.
`-- If an action is an assignment in a format unsupported by Lua (a += 1), convert it into a supported format (a = a + 1)`

File: Classes/ShapeGenerator.lua [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+-- Handler that makes the structure solid. 
+cShapeGenerator.m_SolidHandler = function(a_ShapeGenerator, a_BlockArea, a_BlockPos)
+	return true
+end
+
+
+
+
+
+function cShapeGenerator:new(a_Zero, a_Unit, a_BlockTable, a_Expression)
```
Comment (madmaxoft) [inline Classes/ShapeGenerator.lua]: What are the parameters? Their types? Any return values?

**Comment by NiLSPACE** [comment]
Done

File: Classes/Expression.lua [hunk 5]
```diff
@@ -0,0 +1,253 @@
+
+-- Expression.lua
+
+-- Contains the cExpression class. This allows formulas to be executed safely in an empty environment.
+--[[ Usage example:
+local FormulaString = "data=4; x > y"
+local Expression = cExpression:new(FormulaString)
+
+Expression:AddReturnValue("Comp1")           -- Return the first known comparisons
```
Comment (madmaxoft) [inline Classes/Expression.lua]: Wait, the comparison results can be returned? Why is this not mentioned elsewhere? I'd expect it in the `AddReturnValue` docs and `Compile` docs.

File: Classes/Expression.lua [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	
+	-- Split the formula into actions (For example in "data=5; x<y" data=5 is an action, and x<y is an action.)
+	local Actions = StringSplitAndTrim(self.m_Formula, ";")
+	
+	-- If an action is an assignment in a format unsupported by Lua (a += 1), convert it into a supported format (a = a + 1). 
+	-- If an action is a comparison then give it the name "Comp<nr>"
+	for Idx, Action in ipairs(Actions) do
+		local IsAssignment = true
+		
+		-- If one of the comparison operator's are in the action we can be sure that it's an assignment
```
Comment (madmaxoft) [inline Classes/Expression.lua]: This comment makes little sense.

File: Classes/ShapeGenerator.lua [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+end
+
+
+
+
+
+-- Creates a new cShapeGenerator object.
+-- a_Zero and a_Unit are Vector3f vectors used to calculate a scaled vector3f. The formula will use the scaled vector as x, y and z values.
+-- a_BlockTable are the blocks to make the shape out of.
+-- a_Expression is a cExpression object that compiles the formula into a safe function.
```
Comment (madmaxoft) [inline Classes/ShapeGenerator.lua]: "Compiles", or "is expected to have compiled"

Also, the function should probably check the parameters and assert if they don't match expectations.

**Comment by NiLSPACE** [comment]
Done

**Comment by NiLSPACE** [comment]
**bump**

**Comment by madmaxoft** [comment]
Okay, now I think it's good enough for casual reader to understand most of the code :)

**Pull request merged**